### PR TITLE
Avoid partial construction of finalizable type PinnedRemotableDataScope

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/FatalError.cs [hunk 1]
```diff
@@ -178,6 +178,20 @@ public static bool ReportWithoutCrashUnlessCanceledAndPropagate(Exception except
...
         }
 
+        /// <summary>
+        /// Suppress finalization of an instance when an exception is thrown from the constructor.
+        /// </summary>
+        /// <typeparam name="T">The type being created.</typeparam>
+        /// <param name="instance">The partially-constructed instance for which finalization should be suppressed.</param>
+        /// <returns>Always returns <see langword="false"/> to propagate the exception.</returns>
+        [DebuggerHidden]
+        public static bool SuppressFinalizeForFailedConstruction<T>(T instance)
```
Comment (tmat) [inline src/Compilers/Core/Portable/InternalUtilities/FatalError.cs]: Unused?

File: src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs [hunk 2]
```diff
@@ -62,27 +63,36 @@ internal sealed class PinnedRemotableDataScope : IDisposable
...
         {
-            Contract.ThrowIfNull(solutionChecksum);
-
             _storages = storages;
             _storage = storage;
+            SolutionInfo = solutionInfo;
+        }
 
-            SolutionInfo = new PinnedSolutionInfo(
+        public Workspace Workspace => _storage.SolutionState.Workspace;
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs]: Nit: I'd place these after the factory. It helps to have factories close to constructors.

Comment (sharwell) [reply src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs]: I kept the relative ordering in this file for now (constructors, then properties, then methods).

File: src/Compilers/Core/Portable/InternalUtilities/FatalError.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/FatalError.cs]: Currently yes. I'm still investigating patterns for some other cases and this seems like a well-defined operation should it be needed.

Comment (tmat) [reply src/Compilers/Core/Portable/InternalUtilities/FatalError.cs]: I would rather not add dead code. Let's add it when it's needed.

Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/FatalError.cs]: Removed

**Comment by tmat** [comment]
@sharwell Ok to merge?

**Pull request merged**